### PR TITLE
V2: Compact model loader saver

### PR DESCRIPTION
**Pull request by hwpang**
## Description
This adds features discussed in https://github.com/chemprop/chemprop/pull/585#discussion_r1459315346 and https://github.com/chemprop/chemprop/issues/589.

Essentially, the checkpoint file contains all the information to continue training that are not needed for end users who only wish to use the model for inference.

This PR adds functionality to save and load a model file that contains minimal to run inference: the model architecture and the model state (weights).

This PR adds:
- `save_model`: save the model architecture, the model state, and scalers
- `load_from_file`: load the model file, is also able to load a checkpoint file.
- Unit tests to test above functions

## Relevant issues
https://github.com/chemprop/chemprop/issues/589

### CODE REVIEW

**Review by davidegraff — COMMENTED** [review]
provided that these unit tests pass, my only notes are:
1. we probably want some `load_model` function available somewhere that takes an input `.pt` file and returns the model and our scalers, i.e.,
```python
def load_model(path: PathLike) -> tuple[
    MPNN, list[StandardScaler] | None, StandardScaler | None
]:
    ...
```
2. that we allow for optional input and output scalers when (i) _saving_ the model at the end of `chemprop train` and (ii) _loading_ the model at the start of `chemprop predict`

**Review by davidegraff — DISMISSED** [review]

**Comment by hwpang** [comment]
@davidegraff Thanks for the suggestions! I have added the following
- I added a `load_model` function as you suggested
- I added `load_model` function in `chempro predict` and `save_model` function in `chemprop train`.
  - In `chemprop predict`, if `--checkpoint` ends with `.pkl`, `load_model` is used to load the model file. Otherwise, `load_from_checkpoint` is used. The `input_scalers` and `output_scaler` is not yet connected to the model as I think there are some ongoing discussion in https://github.com/chemprop/chemprop/issues/512 regarding what's the best way to implement output scaling and https://github.com/chemprop/chemprop/issues/613 regarding input_scalers.
  - In `chemprop train`, the last model is saved using the `save_model` function, where the training data scaler is saved as the output scaler.

Additionally, I also added/modified the following
- I moved the `{save,load}_model` to `chemprop.models.utils` to avoid cyclic import
- I added a `load_from_file` function for `MulticomponenetMPNN`
- I added tests for `load_from_file` for `MulticomponenetMPNN`, similar to those for `MPNN`

**Comment by hwpang** [comment]
I see that the build on MacOS is failing currently, although I didn't change anything that should fail it in this PR. @JacksonBurns Do you know what might cause this?

File: chemprop/models/utils.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        },
+        path,
+    )
+
+
+def load_model(path: PathLike) -> tuple[MPNN, list[StandardScaler] | None, StandardScaler | None]:
+    try:
+        model = MulticomponentMPNN.load_from_file(path)
+    except KeyError: # MPNN
+        model = MPNN.load_from_file(path)
```
Comment (davidegraff) [inline chemprop/models/utils.py:24]: I don't like doing something based on try/except logic. Is there some way we can think of embedding the model class into the saved file? Possibly including a `'cls'` key that contains `model.__class__` and just loading that?

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -684,9 +684,11 @@ def main(args):
         results = trainer.test(model, test_loader)[0]
         logger.info(f"Test results: {results}")
 
-    p_model = args.output_dir / "model.pt"
-    torch.save(model.state_dict(), p_model)
-    logger.info(f"model state dict saved to '{p_model}'")
+    p_model = args.output_dir / "model.pkl"
```
Comment (davidegraff) [inline chemprop/cli/train.py:687]:
```suggestion
    p_model = args.output_dir / "model.pt"
```

File: chemprop/cli/predict.py [hunk 3]
```diff
@@ -170,9 +170,15 @@ def main(args):
...
+        if args.checkpoint.endswith(".pkl"):
+            model, input_scalers, output_scaler = load_model(args.checkpoint) # TODO: connect input_scalers and output_scaler to the model
+        else:
+            model = MulticomponentMPNN.load_from_checkpoint(args.checkpoint)
     else:
-        model = MPNN.load_from_checkpoint(args.checkpoint)
+        if args.checkpoint.endswith(".pkl"):
+            model, input_scalers, output_scaler = load_model(args.checkpoint) # TODO: connect input_scalers and output_scaler to the model
+        else:
+            model = MPNN.load_from_checkpoint(args.checkpoint)
```
Comment (davidegraff) [inline chemprop/cli/predict.py:181]: we should have separate concepts of a `checkpoint` and a `saved_model`. The point of making a model state a subset of a checkpoint allows for users to simply supply a checkpoint file as the value of `--model-path`

File: chemprop/models/utils.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/models/utils.py:24]: or you can just supply `multicomponent` as an input argument and that'd probably work better.

**Comment by JacksonBurns** [comment]
> I see that the build on MacOS is failing currently, although I didn't change anything that should fail it in this PR. @JacksonBurns Do you know what might cause this?

It's nothing to do with v2 or chemprop in general, I think.

If you look at the history of test runs on pushes to master: https://github.com/chemprop/chemprop/actions/workflows/tests.yml?query=event%3Apush+branch%3Amaster

you can see that some of the MacOS tests have been hanging and timing out, like this one where all we did was change the license: https://github.com/chemprop/chemprop/actions/runs/6828366831/job/18572355296

though sometimes they work: https://github.com/chemprop/chemprop/actions/runs/7572256483/job/20621648110

TL;DR: Don't worry about it; we might want to look at it closer to release if it keeps happening.

**Review by JacksonBurns — COMMENTED** [review]
Can we do `.ckpt` instead? That's what lightning does by default: https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.ModelCheckpoint.html#modelcheckpoint

**Comment by davidegraff** [comment]
> Can we do `.ckpt` instead? That's what lightning does by default: https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.ModelCheckpoint.html#modelcheckpoint

Do this where? The model state (`.pkl` or `.pt`) is not a valid lightning checkpoint, but a lightning checkpoint is a valid model state

**Comment by JacksonBurns** [comment]
> Do this where? The model state (`.pkl` or `.pt`) is not a valid lightning checkpoint, but a lightning checkpoint is a valid model state

Ah I didn't know they were different types, nevermind!

File: chemprop/models/utils.py (hunk 1 above)
Comment (hwpang) [reply chemprop/models/utils.py:24]: I like having `multicomponent` as the input argument to make it explicit! Will do that.

File: chemprop/cli/predict.py (hunk 3 above)
Comment (hwpang) [reply chemprop/cli/predict.py]: Thanks, that makes sense! I moved the `--checkpoint` and other related arguments from `common.py` to `train.py`, added `--model-path` to `predict.py`, and modified relevant codes to use `--model-path` in `predict.py`.

**Comment by hwpang** [comment]
@davidegraff Thanks for the notes! I have addressed the comments.

**Review by kevingreenman — APPROVED** [review]
